### PR TITLE
#164857216 enable social login

### PR DESCRIPTION
**Pull request by bencyn**
**What does this PR do?**

Enables Social login via Google, Facebook, and Twitter

**Tasks to be completed**

- [x] Configure google, facebook, and twitter API accounts
- [x] Configure social auth package
- [x] Write unit tests

**How can this be manually tested**

- Clone the repo and checkout the `ft/164857216-enable-social-login` branch

- Create and activate a virtual environment
```
$ virtualenv env
$ source env/bin/activate
```

- Install requirements
```
$ pip install -r requirements.txt
```

- Make migrations
```
$ python manage.py makemigrations
```

- Run migrations

```
$ python manage.py migrate
```

- Run the tests
```
python manage.py test
```
- Test via Postman 
  `http://localhost:8000/api/social_auth`
   
   Get Access Token
   [Google Developer](https://developers.google.com/oauthplayground)
     -  Authorize Api
       image
    - Exchange authorization code for tokens 
      image
    - Pass the token to the endpoint as data together with the provider name

```
{
     "provider": "google-oauth2",
     "access_token": ""
}
```
*
    Create facebook App 
    [Facebook Developer](https://developers.facebook.com/apps)
    Use the following details in your settings
    image
    Get Access Token 
    [Facebook Access Tokens](https://developers.facebook.com/tools/accesstoken)
     
    To authorize access to email
    [Authorize email access](https://developers.facebook.com/tools/explorer)
   image

```
{
     "provider": "facebook",
     "access_token": ""
}
```
*
    Create a Twitter App and Get Access Token
    [Twitter Developer](https://developer.twitter.com/en/apps)
    Use the following details 
   image

```
{
     "provider": "twitter",
     "access_token": "",
     "access_token_secret":""
}
```

**Relevant Pivotal tracker stories** 
[#164857216](https://www.pivotaltracker.com/story/show/164857216)

### Screenshots

***Google Login***
image

***Facebook Login*** 
image

***Twitter Login***
image

### CODE REVIEW

**Comment by martinMutuma** [comment]
image
experiencing that error

File: authors/urls.py [hunk 1]
```diff
@@ -15,10 +15,13 @@
...
 
 urlpatterns = [
     url(r'^admin/', admin.site.urls),
-
     url(r'^api/', include('authors.apps.authentication.urls')),
-]
 
+    url(r'^logout/$', auth_views.LogoutView.as_view(), name='logout'),
+    url(r'^oauth/', include('social_django.urls', namespace='social')),
+    url(r'^admin/', admin.site.urls),
```
Comment (Tevinthuku) [inline authors/urls.py]: You have 2 admin urls, you can get rid of 1

**Comment by Tevinthuku** [comment]
Do you have tests for the social login endpoints?

**Comment by Tevinthuku** [comment]
Do you have tests for the social login?

**Comment by Tevinthuku** [comment]
@bencyn how do I get the token from the Google playground ?https://developers.google.com/oauthplayground/

**Comment by Tevinthuku** [comment]
@bencyn is this test passing on your machine
image

I printed out the data and as you can see in the screenshot it says that the access token cannot be null

**Comment by Tevinthuku** [comment]
> @bencyn is this test passing on your machine
> image
> 
> I printed out the data and as you can see in the screenshot it says that the access token cannot be null

I think I know why its failing, maybe you can share the new keys you are using with the team like you did last week

File: authors/apps/authentication/views.py [hunk 2]
```diff
@@ -230,3 +237,98 @@ def list(self, request):
...
+
+        except AuthForbidden as error:
+            return Response({
+                "error": "invalid token",
+                "details": str(error)
+            }, status=status.HTTP_400_BAD_REQUEST)
+
+        if authenticated_user and authenticated_user.is_active:
+            # Check if the user you intend to authenticate is active
+            token = JWTokens.create_token(self, authenticated_user)
```
Comment (Tevinthuku) [inline authors/apps/authentication/views.py]: Why not use the `token` property in the user model so that we can have `authenticated_user.token`

Comment (bencyn) [reply authors/apps/authentication/views.py]: Thanks for the feedback It's done

**Review by Tevinthuku — APPROVED** [review]
This works as expected, you can edit the username so that it doesn't fail in the username validation coming in #13

**Comment by bencyn** [comment]
@Tevinthuku I've cleaned the username field
image

**Review by justMuriithi — APPROVED** [review]

**Pull request merged**